File: longest-prefix/longest_prefix.js
function longestPrefix(words) {
    if(words.length === 0) {
        return "";
    } else if(words.length === 1) {
        return words[0];
    }

    // 2 or more
    let prefix = undefined;
    for(let i=0; i<words.length-1; i++) {
        if(prefix) {
            prefix = cmp(prefix, words[i+1]);
        } else {
            prefix = cmp(words[i], words[i+1]);
        }
    }

    return prefix;
}

function cmp(a, b) {
    let i=0;
    while(i<a.length && i<b.length && a[i] === b[i]) {
        i++;
    }
    return a.substring(0,i);
}

a = ['hello'];
b = [`floor`, `florida`, `fluid`];
c = [`doodle`, `double`, `dave`, `murmur`];
console.log("hello => ",longestPrefix(a));
console.log("fl => ",longestPrefix(b));
console.log("'' => ",longestPrefix(c));

File: longest-prefix/longest_prefix.go
package main

import "fmt"

func longestPrefix(a []string) string {
	if a == nil || len(a) == 0 {
		return ""
	} else if len(a) == 1 {
		return a[0]
	}

	// 2 or more
	lp := cmp(a[0], a[1])
	for i := 2; i < len(a); i++ {
		lp = cmp(lp, a[i])
	}
	return lp
}

func cmp(a, b string) string {
	i := 0
	for ; i < len(a) && i < len(b) && a[i] == b[i]; i++ {
	}
	return a[:i]
}

func main() {
	fmt.Println("Longest Prefix in Go")
	a := []string{}
	b := []string{"hello"}
	c := []string{"doodle", "double", "dave", "maritime"}
	d := []string{"floor", "florida", "fluid"}

	fmt.Printf("'' => %v\n", longestPrefix(a))
	fmt.Printf("hello => %v\n", longestPrefix(b))
	fmt.Printf("'' => %v\n", longestPrefix(c))
	fmt.Printf("fl => %v\n", longestPrefix(d))
}
